src/Microsoft.VisualStudio.ProjectSystem.Managed/Tree/Dependencies/Snapshot/Filters/ImplicitTopLevelDependenciesSnapshotFilter.cs

```diff
@@ -28,43 +28,31 @@ internal sealed class ImplicitTopLevelDependenciesSnapshotFilter : DependenciesS
             IImmutableSet<string>? projectItemSpecs,
             IAddDependencyContext context)
         {
-            // The check for SharedProjectDependency is needed because a SharedProject is not a dependency reference
-            if (!dependency.TopLevel
-                || dependency.Implicit
-                || !dependency.Resolved
-                || !dependency.Flags.Contains(DependencyTreeFlags.GenericDependency)
-                || dependency.Flags.Contains(DependencyTreeFlags.SharedProjectFlags))
+            if (projectItemSpecs != null                                              // must have data
+                && dependency.TopLevel                                                // top-level
+                && !dependency.Implicit                                               // explicit
+                && dependency.Resolved                                                // resolved
+                && dependency.Flags.Contains(DependencyTreeFlags.GenericDependency)   // generic dependency
+                && !dependency.Flags.Contains(DependencyTreeFlags.SharedProjectFlags) // except for shared projects
+                && !projectItemSpecs.Contains(dependency.OriginalItemSpec)            // is not a known item spec
+                && subTreeProviderByProviderType.TryGetValue(dependency.ProviderType, out IProjectDependenciesSubTreeProvider provider)
+                && provider is IProjectDependenciesSubTreeProviderInternal internalProvider)
```

Comment:
This just moves all the precondition checks into one list and adds some comments. The logic is the same.